src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -162,19 +168,22 @@ private static List<JsonConverter> GetDefaultConverters()
             if (converter is JsonConverterFactory factory)
             {
                 converter = factory.GetConverterInternal(typeToConvert, this);
-                // Allow null converters from the factory. This will result in a NotSupportedException later
-                // and with a nice exception that indicates the parent type.
-            }
 
-            if (converter != null)
+                // A factory cannot return null; GetConverterInternal checked for that.
+                Debug.Assert(converter != null);
+            }
+            else
```

Comment:
Hmm these are verifying two different conditions, so I think we should keep the Asserts separate.

Reply:
My point is that you are repeating the same assertion on both if and else branches and the else branch does not contain any other instruction.

so this code:
```cs
if (converter is JsonConverterFactory factory)
{
    converter = factory.GetConverterInternal(typeToConvert, this);

    // A factory cannot return null; GetConverterInternal checked for that.
    Debug.Assert(converter != null);
}
else
{
    // Since the object and IEnumerable converters cover all types, we should have a converter.
    Debug.Assert(converter != null);
}
```

can be reduced to:
```cs
// Allow redirection for generic types or the enum converter.
if (converter is JsonConverterFactory factory)
{
    converter = factory.GetConverterInternal(typeToConvert, this);
}

// Since the object and IEnumerable converters cover all types, we should have a converter.
Debug.Assert(converter != null);
```

Reply:
Yes but if there's an assertion failure we have no way of knowing if the factory returned null or if the code above what you show was incorrect (there was no default converter for the given Type).

Reply:
This would be the equivalent without the `else`:

```cs
Debug.Assert(converter != null);

// Allow redirection for generic types or the enum converter.
if (converter is JsonConverterFactory factory)
{
    converter = factory.GetConverterInternal(typeToConvert, this);
    Debug.Assert(converter != null);
}
```

Reply:
I remove the `else` and moved the Assert to the previous `if` statement.